[Step2- BE 이민형] 회원 로그인

LeeSwallow opened the pull request:
## 특이 사항
1. `JwtAthenticateInterceptor`를 통해 헤더로 들어온 token을 디코딩하고 request attribute에 커스텀 객체(customAuth)에 주요정보를 저장하는 방식으로  구현하였습니다.
2. aop를 활용하여 controller 단에서 수동으로 수행했던 그룹별 validation을 자동화 하였습니다. `@PreAuthorize(UserRole.class)` 를 통해  권한 수준 확인(customAuth에서 가져옴)과 권한수준에 따른 유효성 검증(requestBody를 읽어와 validator를 통해 특정 그룹을 검증) 을 수행하였습니다.
3. 로그인 후 cookie에 httponly로 access-token을 저장하도록 구현하였습니다.이후 `CookieToAttributeInterceptor` 를 통해 cookie에 토큰이 존재한다면 attribute에 토큰을 저장하고 `JwtAthenticateInterceptor`에서 이를 디코딩해서 저장하도록 구현했습니다. 이로서 로그인을 수행한다면 일정 기간 동안 다른 페이지에 접근가능하도록 구현하였습니다.
4. 모든 test를 `RestAssured`를 통해 수행하도록 변경했습니다 .각 test에 대한 abstract 객체를 만들고 최상위 추상 객체에서 `@AutoConfigureRestDocs`  어노테이션을 통해 문서화 작업까지 자동화 하도록 구현했습니다. 추후에 문서화된 api를 swagger를 통해 apidoc server로 배포할 생각입니다.
5. user, userrole, product 스키마를 만들고 user 도메인, userrole 도메인, product 도메인을 각각 구현하였습니다.
6. user는 `/api/users/me` 와 같은 엔드포인트를 제외하고 관리자권한으로만 수행되도록 하였으며, 관리자 권한에 따라 출력되는 결과물도 달리했습니다.(관리자는 password(인코딩됨), roles 를 추가적으로 확인 가능) 또한 권한에 따라 다른 사용자의 email등도  수정 가능하도록 구현하였습니다. 

## 질문 사항
+ 프로젝트가 커지면서 파일 구조를 정리하기 힘들어지기 시작했습니다. 일단 공통적으로 사용하는 패키지들을 `common`에 넣어 정리했는데 어떻게 패키지 구조를 관리해야 할 지 모르겠습니다.

+ viewController에서 오류 사항을 보여 주기 위해(login, create 등의 error-alert에 표시할 메시지 주입) 부득이하게 controller에 validator를 주입하여 사용하였습니다. dto에 대한 검증이므로 controller에서 처리하였습니다만, service로 옮기는게 좋을까요? 애초에 validator를 직접 사용하는게 옳지 않은 설계일까요?

+ ProblemDetail에 대해 알지 못했을 때, ResponseStatusException이 반환하는 에러(RFC 7231)과 유사하게 직렬화하기 위해 ErrorMessageResponse와 builder class 를 구현하였습니다. 
이후에 이를 활용해서 toProblemDetail이라는 mapper 메서드를 만들어 출력하도록 하였는데 새롭게 builder를 구현하는게 좋을까요?

## 관리자 페이지 
### 로그인
![image](https://github.com/user-attachments/assets/ec28471e-b330-42f3-82fc-12420d8b4874)
### 상품목록
![image](https://github.com/user-attachments/assets/6db628d9-cad7-4a4e-95a2-c390f8f23e50)

### 상품상세(수정 버튼을 통해 변경 가능)
![image](https://github.com/user-attachments/assets/825820ad-28b5-4284-a83e-496c7e8f72de)

### 상품 추가
![image](https://github.com/user-attachments/assets/5119243c-6504-4509-8133-f9effa118633)

## 미구현 사항(추가)
- 관리자 페이지에 회원가입은 구현하지 않았습니다(관리자 계정을 회원가입 할 수 있다는 것이 옳지 않다고 생각했습니다)

- 대신 api 를 통해 관리자 권한인 사용자가 관리자인 계정생성과 권한 부여가 가능하도록 구현했습니다.

- 페이지를 통한 user 관리는 시간부족으로 미구현하였습니다.(선택사항이므로 일단 리뷰요청 드렸습니다)

LeeSwallow committed:
feat: RestAssured 를 활용한 테스트 구현

- RestDoc을 추가할 수 있도록 확장성 있는 테스트 구현
- 추후 restdoc -> swaggerUI 로 비지니스 로직을 건드리지 않고 정적인 swagger api doc을 만들 수 있도록 기본 구성

LeeSwallow committed:
refactor: dto에 직접 승인 여부를 넣는 로직 삭제 및 group 설정

LeeSwallow committed:
refactor: Custom한 ErrorMessageResponse에서 ProblemDetail을 추출하는 로직 추가(추후 변경)

Commit:
docs: 2단계 구현 기능 목록 작성 완료

Commit:
feat: DB scheme 변경 완료

Commit:
refactor: 폴더 계층 구조 정리 및 수정

Commit:
feat(entity): User, UserRole 엔티티 구현, Product 수정

- User 엔티티 설계
- UserRole 설계(Enum)
- Product 스키마 수정(owner_id 추가)
- role pk를 role_name 기반으로 수정

Commit:
feat(dao): User, UserRole dao 구현 & 선언 및 구현, 패키지 분리

- UserDao 구현
- UserRoleDao 구현
- 패키지 단위로 분리 후 인터페이스, 구현(Impl) 분리

Commit:
feat(repository): User, UserRole repository 구현 & 선언 및 구현, 패키지 분리

- UserRepository 구현
- UserRoleRpository 구현
- 패키지 단위로 분리 후 인터페이스, 구현(Impl) 분리

Commit:
feat(service): User Service 구현 & 선언 및 구현, 패키지 분리

Commit:
feat(controller): User 관련 CRUD 서비스 구현

- validator를 통한 비밀번호 검증 구현
- dto로 request, response 구조화
- CRUD controller 구현

Commit:
fix(controller): Product 컨트롤러 Response dto 정의 및 수정

Commit:
feat : jwt 관련 의존성 추가

Commit:
refactor : 난잡해진 폴더 구조 정리

- 도메인 별로 나누어지지 않은 패키지(validator, util, model, aop 등)를 공통 폴더로 이동
- 사용하지 않는 DBServerException 삭제(모두 spring 표준 Exception 대체하였음)

Commit:
feat: jwtToken 생성및 password 인코딩을 위한 유틸리티 구현 완료

Commit:
feat: 인터셉터 추가 및 경로 설정 완료

Commit:
fix: Role의 Collection을 Set으로 변경

Commit:
feat: 인증(로그인, 회원가입) 관련 서비스 구현 완료

Commit:
fix: 인증 에러 handler 추가 및 에러 출력 양식 `ProblemDetail`로 변경

- Interceptor에서 인증 오류(AuthenticationException)를 던지도록 수정
- Interceptor에서 attribute에 저장되는 auth에서 가장 우선순위가 높은 권한을 저장하도록 수정
- ErrorMessageResponse에 toProblemDetail Mapper를 추가해 ProblemDetail을 출력하도록 변경

Commit:
feat: ProductController에서 CustomAuth를 활용하도록 도입

Commit:
feat: Aop를 통해 Role을 Controller 수행 이전에 전처리하도록 수정

Commit:
fix: admin이 아닐 때, 자신이 생성한 product(owner가 자신인) 만 수정 가능하도록 수정

Commit:
fix: users scheme email 필드에 unique 추가
-

Commit:
feat: RoleRepository에 sync(매개변수로 받은 roles에 맞게 table 수정) 하는 기능 추가

Commit:
fix: service에서 User entity를 반환할 때 role을 로드하도록 수정

- 모든 저장, 수정 과정에서 role의 수정이나 추가를 요청받았을 때 적절하게 추가할 수 있도록 수정
- user에 대해 요청받았을때 role도 반환하도록 수정

Commit:
feat: UserController 권한 수정 및 ROLE_USER를 위한 endpoint 추가

- /api/users/me 를 통해 ROLE_USER의 경우에도 자신의 정보를 수정 및 조회 삭제 할 수록 구현
- ROLE_ADMIN일 때 모든 사용자의 정보를 roles와 인코딩된 비밀번호를 포함해 조회할 수 있도록 수정

Commit:
fix: Product Create 관련 에러 수정

- dao에서 ownerId를 저장하지 않는 오류 수정
- service에서 ownerId를 user entity에 주입하지 않는 오류 수정

Commit:
fix: error handling 관련 논리 오류 수정

- userService에서 특정 id를 찾을 수 없을 때 404에러를 던지도록 수정
- stackTrace가 null일 때 직렬화되지 않도록 수정
- RoleRepositoryImpl에서 동기화 성공 시 에러를 던지는 오류 수정

Commit:
feat: 모든 endpoint에 대한 e2e test(integration test) 추가

Commit:
feat: thymeleaf 기초 구조 설계

Commit:
fix: forbidden과 unauthorize 확실히 분리

- aop에서 권한 부족시 forbidden 인증 오류시 unauthorize 에러 반환하도록 수정
- unauthorize를 핸들링할 때, www-authenicate 헤더 추가

Commit:
fix: forbidden과 unauthorize 확실히 분리

- aop에서 권한 부족시 forbidden 인증 오류시 unauthorize 에러 반환하도록 수정
- unauthorize를 핸들링할 때, www-authenicate 헤더 추가

Commit:
fix: 문서 레이아웃 수정

LeeSwallow committed:
feat: cookie를 통해 접근 권한을 획득가능하도록 수정

- login시 cookie에 토큰을 저장하고 CookieToAttributeInterceptor에서 이를 불러와 attribute에 불러오도록 수정
- login을 거치지 않는다면 dashboard와 login만을 접근 가능하도록 구현

LeeSwallow committed:
docs: API 명세서 작성 및 구현 기능 목록 수정

riroan commented on src/main/java/gift/common/util/PasswordEncoder.java:
```diff
@@ -0,0 +1,44 @@
+        // 객체 생성을 방지하기 위한 private 생성자
+    }
+
+    public static String encode(String password) {
```
비밀번호 단방향 암호화도 구현하셨군요 👍

riroan commented on src/main/java/gift/common/util/PasswordEncoder.java:
```diff
@@ -0,0 +1,44 @@
+            String hashedRawPassword = encode(rawPassword);
+            return hashedRawPassword.equals(encodedPassword);
+        } catch (Exception e) {
+            throw new RuntimeException("비밀번호 비교 중 오류가 발생했습니다.", e);
```
요기도 Custom Exception을 사용하면 어떨까요?

LeeSwallow replied:
인코딩 알고리즘이 지원하지 않을 때 오류가 발생할 수 있습니다. 이 에러가 발생하면 서버에 지속적으로 치명적인 문제가 발생할 수 있습니다. `CriticalServerException`을 만들고 핸들링 되면 문제를 알린뒤 서버를 종료하는 방식으로 핸들링해도 되려나요

riroan commented on src/main/java/gift/common/util/PasswordEncoder.java:
```diff
@@ -0,0 +1,44 @@
+        try {
+            String hashedRawPassword = encode(rawPassword);
+            return hashedRawPassword.equals(encodedPassword);
+        } catch (Exception e) {
```
Bare exception 보다는 구체적인 exception을 명시하면 좋을 것 같아요. 처리할 exception 종류가 너무 많으면 Exception 으로 써도 되고요!

riroan commented on src/main/java/gift/common/util/TokenProvider.java:
```diff
@@ -0,0 +1,115 @@
+        }
+        if (expiration == null || expiration <= 0) {
+            throw new Exception("설정을 통해 JWT 만료 기간을 제공해야 합니다.\njwt.expiration=your_expiration_time_in_seconds");
+        }
```
여기는 500 에러를 의도한 곳일까요?

LeeSwallow replied:
필수적인 jwt secret key가 누락되어 발생되는 에러이므로 500이 맞다고 판단했습니다. 이 예외지점 또한 서버 로직 전반에 치명적인 에러를 가져올 수 있고 그렇다고 기본값을 설정하는 건 옳지 않다고 생각합니다. 
이 에러도 `CriticalServerException`으로 설정하고자 합니다.

riroan commented on src/main/java/gift/common/util/TokenProvider.java:
```diff
@@ -0,0 +1,115 @@
+            log.info("만료된 JWT 토큰입니다.");
+        } catch (Exception e) {
+            log.info("유효하지 않은 JWT 토큰입니다.");
+        }
```
여기는 log level을 warning 쯤으로 올리면 어떨까요?

riroan commented on src/main/java/gift/controller/view/DefaultViewController.java:
```diff
@@ -0,0 +1,109 @@
+        }
+        if (tokenProvider.getAuthentication(token).role() != UserRole.ROLE_ADMIN) {
+            throw new IllegalArgumentException("관리자 권한이 필요합니다.");
+        }
```
권한이 맞지 않는 거라서 IllegalArgumentException은 좀 어색한 것 같아요 다른 Exception을 써보면 어떨까요?

LeeSwallow replied:
error를 던지는 목적이 `adminLogin` 로직 내부에서 이를 catch 해서 에러 메시지를 추출 및 view에 출력(html의 error_alert에 출력)할 목적으로 사용하여 아무 에러나 던졌던 것 같습니다.
`AccessDeniedException`으로 수정하겠습니다.

riroan commented on src/main/java/gift/repository/role/RoleRepositoryImpl.java:
```diff
@@ -0,0 +1,74 @@
+    public Boolean save(Long userId, UserRole role) {
+        if (exists(userId, role)) {
+            throw new IllegalArgumentException("이미 존재하는 사용자 역할입니다.");
+        }
```
이미 존재하는 사용자는 409 Conflict로 처리할 수 있기 때문에 Exception을 분리하면 좋을 것 같아요!

riroan commented on src/main/java/gift/repository/role/RoleRepositoryImpl.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Override
+    @Transactional
+    public Set<UserRole> sync(Long userId, Set<UserRole> roles) {
```
여기에서 중요한 건 아니지만 비트집합을 사용하면 더 효율적으로 나타낼 수 있어요!

riroan commented on src/main/java/gift/service/product/ProductServiceImpl.java:
```diff
@@ -21,6 +23,29 @@ public ProductServiceImpl(ProductRepository productRepository) {
+            log.error("상품 소유자 ID가 인증된 사용자 ID와 일치하지 않습니다. 소유자 ID: {}, 인증된 사용자 ID: {}",
+                      product.getOwnerId(), auth.userId());
+            throw new IllegalArgumentException("상품 소유자 ID가 인증된 사용자 ID와 일치하지 않습니다.");
+        }
```
이런 경우 403을 써보면 어떨까요?

riroan commented on src/test/java/gift/auth/AuthTest.java:
```diff
@@ -0,0 +1,108 @@
+                .body("detail", notNullValue())
+                .body("validationErrors", Matchers.hasSize(1));
+    }
+
```
비밀번호 조건에 맞지 않는 케이스 몇 개를 추가하면 어떨까요?
예를 들면 특수문자가 없는 케이스, 너무 짧은 케이스 등이 있을 것 같아요!

riroan reviewed (changes requested):
안녕하세요 민형님~
미션을 너무너무 잘 수행해주셨어요! 프로덕션에 나갈 코드라고해도 손색이 없을 정도였어요. 💯 
다양한 스프링의 기능들을 적절하게 사용하시고 역할별로 클래스를 나눈 게 눈에 띄었어요!
미션 수행하시느라 너무 고생 많았습니다~

몇 가지 코멘트를 남겨드렸는데 확인 후 리뷰요청 부탁드려요 🙏 
궁금한 점이나 논의하고 싶은 점은 PR comment나 DM 으로 남겨주세요!

riroan commented:
질문에 대한 답변 드려요~
1. 저같은 경우는 공통 패키지들은 프로젝트 최상단에 util, common등의 폴더를 만들어서 관리하는 편이에요. 프로젝트 어디든 쓰일 수 있기 때문이에요. 그리고 패키지 구조 관리는 크게 두 가지로 `도메인 별로 분리`와 `레이어 별로 분리`가 있을 것 같아요. 둘 중 어떤 걸 사용해도 괜찮지만 섞어서 쓰는 것은 지양해야한다고 생각합니다! 민형님같은 경우는 레이어별로 잘 나누신 것 같아요 💯 

2. 저는 controller에서는 최소한의 분기만 하고 로직을 Service로 넘겨야한다고 생각해요. 레이어간 책임을 봤을 때 Controller는 요청을 받아서 하위 레이어로 넘기는 역할, Serice는 비즈니스로직을 수행하는 역할을 가지기 때문이에요. 그래서 최대한 validator는 DTO 필드 validatation이나 Service에서 하는 게 좋을 것 같아요!

3. 이 부분에 정답은 없는 것 같아요! 두 방법 다 목적을 이루기 위한 수단이기 때문에 편한 방법으로 구현하시면 좋을 것 같습니다 ㅎㅎ

Commit:
refactor : Custom 예외로 util 에서 발생하는 에러에 대한 핸들링 추가 및 수정

Commit:
refactor : 도메인에서 발생하는 에러 처리 세분화 및 수정

- 이미 존재하는 사용자이거나, 존재하는 이메일에 대한 에러는 409 Conflict를 반환하도록 수정했습니다.
-
- deleteById 시 (사용자 존재 확인 -> 삭제)를 transactional 하게 수행합니다. 이때 사용자가 삭제되지 않는다면 illegalStateException을 통해 500을 던지도록 했습니다.

- 500 에러에 한에서 log를 추가했습니다.

Commit:
refactor: jwt 인증 실패 관련 log level 증가

Commit:
refactor: 로그인 불일치 401 unathorize로 변경

- AuthenticateException -> UnathorizedException으로 변경
- 존재하지 않는 email로 인한 실패도 401으로 반환하도록 수정

Commit:
refactor: 전반적인 테스트 케이스 보강 및 수정

- AuthTest에 validation 케이스 추가
- Product 엔드포인 관련 테스트 케이스 보강
- User 엔드포인트 관련 테스트 케이스 보강
- 누락된 UserDeleteTest 추가

Commit:
fix: 잘못된 테스트 계정 정보 수정

LeeSwallow commented:
## 특이 사항
1. `CriticalServerException` 핸들링 과정에서 서버를 강제 종료 하기보단 500 에러를 반환하도록 했습니다.
2. `Authentication`(java 기본 exception)에서 `UnathorizeException`(custom)으로 변경했습니다.
    - checked 에러이며 spring에서 관리되는 에러가 아니므로 일단 직접 만들었습니다.
    - `spring.security`을 사용하지 않는 것을 권장하였기에 따로 만들었지만 추후에 도입된다면 security에 내장된 에러를 사용하는것이 좋지 싶습니다.
3. Test 전, 후처리를 추가적으로 보완하고 케이스를 추가했습니다. 추상 클래스에서 테스트 객체 생성과 삭제를(`@BeforeEach`, `@AfterEach`를 통해) 전담해 테스트가 실제 프로젝트에 영향을 주지 못하도록 했습니다.
    - `@Profile` 설정으로 배포환경(prod)과 테스트 환경(test)을 분리할 수 있도록 설정했습니다.

Commit:
feat: PostContruct를 통해 서버 시작 전에 치명적인 문제들에 대해 미리 검증할 수 있도록 구성

LeeSwallow commented:
## 추가 보완

+ 서버시작 전에 발생할 수 있는 문제를 미리 검증하고 서버 시작자체를 막을 수 있도록 `ServerStartupVerifier` 클래스를 도입했습니다. `@PostContruct`를 통해  서버 시작 후 bean에 등록되자 마자 필요한 검증들을  미리 수행합니다. 
`Component`에 이름을 설정하고 `DependsOn`으로 우선순위를 설정하여 검증 대상 객체로 인해 충돌이 나는 문제를 해결했습니다.

Commit:
feat: 추후 지원을 위해 SortBy Record 완성

riroan commented on src/main/java/gift/common/util/PasswordEncoder.java:
```diff
@@ -1,22 +1,18 @@
-    private PasswordEncoder() {
-        // 객체 생성을 방지하기 위한 private 생성자
-    }
+    private static final String ALGORITHM = "SHA-256";
```
지금 꼭 반영할 필요는 없지만 이런 값도 conf (properties)에 넣으면 좋을 것 같아요. 나중에 algorithm 이 바뀔 때 코드에 있으면 코드수정이 필요하지만 conf에 있으면 conf 만 바꾸고 app을 restart 하면 된다는 장점이 있어요!

riroan reviewed (approved):
안녕하세요 민형님~
피드백 반영하신 것들 잘 확인했습니다!
2단계 미션 진행하시느라 고생하셨습니다 💯 
이번 미션은 여기서 머지하도록 하겠습니다. 다음 미션도 파이팅 🔥
